### PR TITLE
Vectorize Python implementation of Poisson solutions

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Not used in spectre (so far) but useful for importing the solutions from elsewhere.

### Types of changes:

- [x] Improvement

### Component:

- [x] Code

### Code review checklist

- [x] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/ProductOfSinusoids.py [hunk 1]
```diff
@@ -5,22 +5,36 @@
...
 def source(x, wave_numbers):
-    x, wave_numbers = np.asarray(x), np.asarray(wave_numbers)
+    x, wave_numbers = parse_arguments(x, wave_numbers)
     return np.sum(wave_numbers**2) * field(x, wave_numbers)
 
 def auxiliary_source(x, wave_numbers):
-    return np.zeros(x.shape)
-# End functions for testing ProductOfSinusoids.cpp
+    x, wave_numbers = parse_arguments(x, wave_numbers)
+    return np.zeros(x.shape)# End functions for testing ProductOfSinusoids.cpp
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/ProductOfSinusoids.py]: Why not have the comment be on its own line anymore?

**Review by nilsdeppe — COMMENTED** [review]
I personally don't like the term `vectorize` outside of the context of using vector registers on the CPU. However, I don't know enough about python to know if that's the goal of this commit or if it's just to support numpy array math. In the latter case I don't think we should use the term 'vectorize' since it already has an established different term in HPC.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/ProductOfSinusoids.py (hunk 1 above)
Comment (nilsvu) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/ProductOfSinusoids.py]: was not intentional :) no idea what went wrong

**Comment by nilsvu** [comment]
@nilsdeppe The term has a particular meaning in the context of numpy math: https://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.vectorize.html
Therefore I think it's correct to use it here. But it's only the commit message so I don't care very much about it.

**Comment by osheamonn** [comment]
Could you elaborate on the rationale for this change, so I've a better idea of the problem it's trying to solve? If it's not used I don't see why it's needed, and I don't really understand what you mean by importing functions from elsewhere.

**Comment by nilsvu** [comment]
@osheamonn I import these analytic solutions from my Python prototyping code (that I use for experimenting with DG formulations before implementing them in spectre). It helps to work with the code that's already in spectre instead of duplicating it. The change in this PR makes the functions work for when the coordinates are arrays. This is not currently needed for the spectre testing code, but is generally expected behaviour in Python and useful in my prototyping code since it makes the functions more flexible in what inputs they can handle.

**Comment by nilsvu** [comment]
does this answer your question @osheamonn? then this should be ready

**Comment by osheamonn** [comment]
So I'm still a bit hesitant about the change. I can't think of a situation in the spectre testing environment where we'd want the co-ordinates to be arrays, and so I'd rather defer implementing that change until it's clear that it's needed. And I appreciate it that it's probably useful for you to have the same code in your prototyping code as in spectre, I don't think that's enough of a reason to have it merged into spectre.

**Comment by mar-celine** [comment]
I think this will be good to have when we have something that needs it, but as @osheamonn said, until we do, we should hold off on putting it into SpECTRE for now.

**Comment by nilsvu** [comment]
I think this only makes the Python code better, not worse. There is no downside in merging it. If you are opposed to that, though, we can close this PR

**Comment by osheamonn** [comment]
I think it makes the python code more generally applicable, just to a use case we don't require yet. And I think there is a readability and complexity cost to the generalisation (required reshapes, extra control flow with expand_dims), particularly for those that aren't as well versed in python/NumPy.

**Comment by nilsvu** [comment]
Closing this in response to the discussion in the meeting today. The agreement was that if we will want to expose this to a Python module at some point it should be moved out of `tests` to the appropriate place in `src`. For the testing code we prefer minimal working versions of these functions that are easier to understand.

**Pull request closed without merging**